waspc/src/Wasp/Generator/Monad.hs

```diff
@@ -0,0 +1,78 @@
+  ( Generator,
+    GeneratorError (..),
+    GeneratorWarning (..),
+    catchGeneratorError,
+    logAndThrowGeneratorError,
+    logGeneratorWarning,
+    runGenerator,
+  )
+where
+
+import Control.Arrow (left)
+import Control.Monad.Except (ExceptT, MonadError (throwError), runExceptT)
+import qualified Control.Monad.Except as MonadExcept
+import Control.Monad.Identity (Identity (runIdentity))
+import Control.Monad.State (MonadState, StateT (runStateT), modify)
+import Data.List.NonEmpty (NonEmpty, fromList)
+
+-- | Generator is a monad transformer stack where we abstract away the underlying
+-- concrete monad transformers with the helper functions below. This will allow us
+-- to refactor and add more transformers (or swap them) without any caller changes.
+--
+-- The outer Either layer represents the error that halted generation. Every error is fatal.
+-- There may optionally be additional errors or non-fatal warnings logged in the State.
+newtype Generator a = Generator
+  { _runGenerator :: ExceptT GeneratorError (StateT GeneratorState Identity) a
+  }
+  deriving
+    ( Functor,
+      Applicative,
+      Monad,
+      MonadState GeneratorState,
+      MonadError GeneratorError
+    )
+
+data GeneratorState = GeneratorState
+  { warnings :: [GeneratorWarning],
+    errors :: [GeneratorError]
+  }
+
+data GeneratorError = GenericGeneratorError String
+
+instance Show GeneratorError where
+  show (GenericGeneratorError e) = e
+
+data GeneratorWarning = GenericGeneratorWarning String
+
+instance Show GeneratorWarning where
+  show (GenericGeneratorWarning e) = e
+
+-- Runs the generator and either returns a result, or a list of 1 or more errors.
+-- Even if successful there may be warnings, so they are always included.
+runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
+runGenerator generator =
+  let (result, state') = runIdentity $ runStateT (runExceptT (_runGenerator generator)) state
+   in (warnings state', left (\_ -> fromList $ errors state') result)
+  where
+    state = GeneratorState {warnings = [], errors = []}
+
+-- This logs a warning but does not short circuit the computation.
+logGeneratorWarning :: GeneratorWarning -> Generator ()
+logGeneratorWarning w = modify $ \GeneratorState {errors = errors', warnings = warnings'} ->
+  GeneratorState {errors = errors', warnings = w : warnings'}
+
+-- This logs an error and does throw, thus short-circuiting the computation unless caught.
+logAndThrowGeneratorError :: GeneratorError -> Generator a
+logAndThrowGeneratorError e = logGeneratorError >> throwError e
+  where
+    logGeneratorError :: Generator ()
+    logGeneratorError = modify $ \GeneratorState {errors = errors', warnings = warnings'} ->
+      GeneratorState {errors = e : errors', warnings = warnings'}
+
+-- This stops the short-circuiting from above, if ever desired. Note, the monad behavior
+-- should ensure that any logged error results in failure, however.
```

Comment:
What do you mean by `the monad behavior
-- should ensure that any logged error results in failure, however`?

I think they can't avoid it -> once error has been logged, it will for sure be treated as a failure, there is no way to get that error out of there, which is great. Aha but that is exactly what you said, isn't it -> you said even if you catch it, the whole computation will still be considered a failure, due to how monad internally works. Ok got it now, all good!

Reply:
Here I was trying to convey a responsibility for the the caller of that function. For example, at any point one could do something like:

```hakell
foo `catchGeneratorError` (\_ -> return [])
```

Where if foo threw, it gets logged but they swallow it and do not re-throw anything. This may allow for the whole process to succeed, as the `EitherT` wont be triggered. This is a "recovery" mechanism where some errors are not fatal (which we agreed we don't have a great use for atm, but this still allows for that in the future).

So I can probably just remove the Note since it just reinforces our intended use of `any throws` will be fatal, but is more confusing than helpful I think.

Reply:
Aha, but even if they do that, if they "swallow" the error, the errors is still logged in the monad state, right? And there is no way they can remove that! So once they call runGenerator, that error will be in the result of it, and no file drafts will come out, therefore making it a fatal error. So they actually can't mess it up, even if they catch it and ignore it -> once it is logged, it becomes a fatal error, due to the type and implementation of runGenerator, and the fact that there is no way to modify list of logged errors besides adding a new one.

Reply:
Well, not quite (in my testing at least). If they catch it, it does get logged into state, but the generator result can now be whatever they pass back. The only way the ErrorT is an error as a result is if they throw it all the way to the top. I don't know that we can change that behavior.

Example: https://replit.com/@smcnc/LopsidedPoliteSymbol#Main.hs

Reply:
Well, I suppose we _could_ in `runGenerator` override the default `ExceptT` Monad behavior and check if there are any errors in the list, and if so, we can just take the first and force it to be `Left` in the return? Not sure if we want to do that or not though, as that feels limiting for future use where some errors may be recoverable and may be confusing for the overall use of catch.

I'll hold off on merging until I hear back (no rush, can with until Monday). My initial gut reaction says let's keep existing behavior but open to feedback. Thanks

Reply:
Ah I see what you mean now, sorry I got confused a bit!

You are right, it behaves as you described.

I would actually be tempted to do this other approach that you described above, where we return the list of errors if (Left latestError) is returned, or if (Right fds) is returned but there are errors in the state -> so in both of those cases we could return errors. Only if there is (Right fds) and no errors logged in the state, then we would forward file drafts.

However, I don't feel this is limiting for the future use. Ok let's see, so what I am aiming for is:
- Once error is thrown, there is no "true" recovery -> that generator, and all the generators using/calling it, failed. I don't think they should ever be able to recover. Otherwise it would be a warning, not an error. But maybe this is too strict? Maybe we would like to throw some errors that the thrower can actually work around? Hmmmm. But if we will want to do it, we can always refactor to support that.
- The type of recovery we do support is recovery for the purpose of collecting additional errors, not turning failed computation into a successful one. So generator can catch an error form sub-generator and decide to continue calling other parts of generator in order to collect more errors, but fully knowing that it can't really recover. This is how it is usually done in Parsing -> once you get an error, you know that something is wrong, but you want parser to recover so you can collect more errors, so that user doesn't have to fix error by error but can instead see more of them at the same time. However that parsing is still failed, for sure. This is how it was done in book "Crafting Interpreters" and it made a lot of sense to me.

So I guess the whole question is: do we want to allow "real" recovery, or only recovery for the purpose of collecting more errors? I am voting for the second, because I think there is a fair chance we will never need the first option, and by being a bit stricter like this, we are sure that we can't by mistake catch and e.g. not rethrow an error -> now I can be sure in Generator when I throw an error that the whole computation will fail, regardless if somebody tries to catch it or now, and that feels pretty important to me. We could in future add smth like `throwRecoverableError` if we really want, and then we can handle those differently, for example put them in another list and then threat that list differently -> so there will be ways to add that if needed -> but I would rather that we stay strict and safe for now.

I hope I explained this well, let me know how this sounds / if it makes sense. As I said, I saw this first in that book, being used for Parser, and it makes perfect sense to me so I am trying to push it here also.

Reply:
Alrighty, we can do something like this:

```haskell
runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
runGenerator generator =
  let (eitherResult, state') = runIdentity $ runStateT (runExceptT (_runGenerator generator)) state
   in (warnings state', finalResult eitherResult (errors state'))
  where
    state = GeneratorState {warnings = [], errors = []}
    finalResult eitherResult loggedErrors = case eitherResult of
      Left _ -> Left $ fromList loggedErrors
      Right result
        | null loggedErrors -> Right result
        | otherwise -> Left $ fromList loggedErrors
```

And results in the behavior we'd like: https://replit.com/@smcnc/BronzeGoldenrodBrowsers#Main.hs

Sorry it took me a while to get on the same page for that unrecoverable error behavior. 😅 It should have been clear in prior comment discussions, but for some reason from early on I was thinking we wanted recoverability and it stuck. But we are now on the same page!!! 👍🏻 ref: https://github.com/wasp-lang/wasp/commit/de7dca49e522cd4fe4f022de33f22aef2b319c83

Reply:
Perhaps even a bit clearer this way:

```haskell
runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
runGenerator generator =
  let (eitherResult, state') = runIdentity $ runStateT (runExceptT (_runGenerator generator)) state
   in (warnings state', loggedErrorsOrResult (eitherResult, errors state'))
  where
    state = GeneratorState {warnings = [], errors = []}
    loggedErrorsOrResult (Left _, []) = error "Generator produced error, but had empty log!"
    loggedErrorsOrResult (Left _, loggedErrors) = Left $ fromList loggedErrors
    loggedErrorsOrResult (Right result, []) = Right result
    loggedErrorsOrResult (Right _, loggedErrors) = Left $ fromList loggedErrors
```

Reply:
Hey no worries, it is a complex thing to discuss and I also wasn't straight with my own thoughts at the every moment, but we managed to figure it out at the end and that is great :)!

```hs
runGenerator :: Generator a -> ([GeneratorWarning], Either (NonEmpty GeneratorError) a)
runGenerator generator =
  let (errorOrResult, finalState) = runIdentity $ runStateT (runExceptT (_runGenerator generator)) initialState
   in (warnings finalState, loggedErrorsOrResult (eitherResult, errors finalState))
  where
    initialState = GeneratorState {warnings = [], errors = []}
    
    loggedErrorsOrResult (Right result, []) = Right result
    loggedErrorsOrResult (Left _, []) = error "Generator produced error, but had empty log - this should never happen!"
    loggedErrorsOrResult (_, loggedErrors) = Left $ fromList loggedErrors
```

I did couple of small changes here, pick those that you like and drop those you don't!

Clarifications:
- When something is Either Foo Bar I like naming it `fooOrBar` (unless I have a better name), since it seems closest to what it is.
- `state` and `state'` is nice because it shows that `state'` comes after/from `state`, however `state` is really initial state so I thought naming it that way gives us more information, indicates that it is simple, empty or has default values. So it is not some state, it is initial state.
- I added "this should never happen" so that in the future it is very clear to whomever is reading this piece of code that this is not a badly handled error, but that this should really never happen, and that is why it is error (and not exception). But maybe that is redundant.
- I merged last two clauses of that function since they have the same logic. I think it is great that you separated that case with invalid state and threw error for it, I totally forgot about that.

Reply:
Hey @Martinsos thanks! All of those suggestions make perfect sense and clean up/clarify things well, so I will use them all. :) I even had initial/final state but went back to what I had before, but your explanation is spot on and it means my gut is getting a bit more aligned I think with where it should be heading.

Thanks again for all the collaboration, I am really happy where this Generator monad ended up! 🥳